### PR TITLE
TR-699: Feature flagged links to snippets

### PR DESCRIPTION
**Pull request by bkemper**

### CODE REVIEW

File: src/components/contentEditor/AceWrapper.js [hunk 1]
```diff
@@ -17,6 +17,11 @@ const AceWrapper = ({
   meta: { error, active, submitFailed }
 }) => (
   <div>
+    {submitFailed && !active && error && (
+      <div className={styles.ErrorWrapper}>
+        <Error error={error} />
+      </div>
+    )}
```
Comment (bkemper) [inline src/components/contentEditor/AceWrapper.js]: Let me know if this is too annoying.

File: src/pages/templates/components/ImportSnippetPanel.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+          component: ExternalLink,
+          content: 'Learn More',
+          to: LINKS.SNIPPET_SUBSTITUTION_REFERENCE
+        }]}
+      >
+        <Panel.Section>
+          <p>
+            Find a snippet, copy the code, and paste it in your template.
+          </p>
+          <Typeahead
```
Comment (bkemper) [inline src/pages/templates/components/ImportSnippetPanel.js]: I need to replace this with a link to the create snippet page when snippets.length === 0.

File: src/config/routes.js [hunk 3]
```diff
@@ -307,23 +307,23 @@ const routes = [
   {
     path: '/snippets',
     component: snippets.ListPage,
-    condition: hasGrants('templates/view'),
+    condition: all(hasGrants('templates/view'), hasUiOption('snippets')),
```
Comment (ischer) [inline src/config/routes.js]: I still seem to be able to access these on UAT with CID 101 (appteam), even with options.ui.snippets === false

File: src/pages/templates/components/ImportSnippetPanel.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import React from 'react';
+import { Button, Label, Panel } from '@sparkpost/matchbox';
+import CopyField from 'src/components/copyField';
+import ExternalLink from 'src/components/externalLink';
+import PanelLoading from 'src/components/panelLoading';
+import { Typeahead, TypeaheadItem } from 'src/components/typeahead/Typeahead';
+import { LINKS } from 'src/constants';
+import { slugToFriendly } from 'src/helpers/string';
+
+export default class ImportSnippetPanel extends React.Component {
```
Comment (ischer) [inline src/pages/templates/components/ImportSnippetPanel.js]: i get an error on this panel when i click into the typeahead, and then click off without selecting a snippet 
```
Uncaught TypeError: Cannot read property 'id' of null
    at Object.itemToString (ImportSnippetPanel.js:45)
    at downshift.esm.js:2118
    at Downshift.<anonymous> (downshift.esm.js:1723)
    at getStateFromUpdate (react-dom.development.js:6400)
    at processUpdateQueue (react-dom.development.js:6489)
    ...
```

File: src/pages/templates/EditPage.js [hunk 5]
```diff
@@ -130,7 +131,7 @@ export default class EditPage extends Component {
             <Form name={formName} subaccountId={subaccountId} readOnly={!canModify} />
           </Grid.Column>
           <Grid.Column xs={12} lg={8}>
-            <ContentEditor readOnly={!canModify} />
+            <ContentEditor readOnly={!canModify} action={canModify && <ImportSnippetLink />} />
```
Comment (ischer) [inline src/pages/templates/EditPage.js]: should this be hidden behind the ui option flag?

File: src/config/navItems.js [hunk 6]
```diff
@@ -113,5 +113,11 @@ export default [
     labs: true,
     divider: true,
     icon: NotificationsActive
+  },
+  {
+    label: 'Snippets',
```
Comment (ischer) [inline src/config/navItems.js]: should this be hidden if the ui option is not set?

File: src/pages/templates/components/ImportSnippetPanel.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            onChange={this.handleChange}
+            placeholder="Type to search..."
+            renderItem={({ id, name }) => (
+              <TypeaheadItem id={id} label={name || slugToFriendly(id)} />
+            )}
+            results={snippets}
+          />
+          <Label>Snippet Code</Label>
+          <CopyField
+            value={`{{ render_snippet('${this.state.snippetId || 'example-id'}') }}`}
```
Comment (ischer) [inline src/pages/templates/components/ImportSnippetPanel.js]: minor thing - it looks like the snippet code reads 
```{{ render_snippet('example-id') }}``` on the template edit page, but it looks like this on the snippet page: ```{{ render_snippet("bob-evans") }}```

File: src/pages/templates/CreatePage.js [hunk 8]
```diff
@@ -53,7 +54,7 @@ export default class CreatePage extends Component {
             <Form newTemplate name={formName} subaccountId={subaccountId}/>
           </Grid.Column>
           <Grid.Column xs={12} lg={8}>
-            <ContentEditor />
+            <ContentEditor action={<ImportSnippetLink />} />
```
Comment (ischer) [inline src/pages/templates/CreatePage.js]: should this import link be hidden behind the ui option flag?

File: src/components/contentEditor/AceWrapper.js (hunk 1 above)
Comment (ischer) [reply src/components/contentEditor/AceWrapper.js]: how can i get this error to show without tweaking code to force an error to be thrown?

Comment (bkemper) [reply src/components/contentEditor/AceWrapper.js]: If you leave both HTML and text empty then attempt to save.

File: src/pages/templates/EditPage.js (hunk 5 above)
Comment (bkemper) [reply src/pages/templates/EditPage.js]: It should.  Good catch.

File: src/config/routes.js (hunk 3 above)
Comment (bkemper) [reply src/config/routes.js]: The condition helper just checks for the existence of the ui option and ignores its value.  I think think this is a little misleading.  I started a thread in Slack.

Comment (ischer) [reply src/config/routes.js]: just saw the thread - this sounds fair, but agree that it's a tad confusing. i'll see if i can just fiddle with the options in dynamo to test.

File: src/pages/templates/components/ImportSnippetPanel.js (hunk 4 above)
Comment (bkemper) [reply src/pages/templates/components/ImportSnippetPanel.js]: Addressed... fe3053e38

File: src/pages/templates/EditPage.js (hunk 5 above)
Comment (bkemper) [reply src/pages/templates/EditPage.js]: Addressed... e4cb9cfd8

File: src/config/navItems.js (hunk 6 above)
Comment (bkemper) [reply src/config/navItems.js]: Yep.

**Comment by bkemper** [comment]
@ischer I addressed all feedback.  Ready for final review.

File: src/pages/templates/components/ImportSnippetPanel.js (hunk 2 above)
Comment (ischer) [reply src/pages/templates/components/ImportSnippetPanel.js]: following up here to see if you wanted to make this change or if you want to defer

File: src/pages/templates/components/ImportSnippetLink.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        <Modal open={this.state.isOpen} onClose={this.handleClose}>
+          <ImportSnippetPanel onClose={this.handleClose} />
+        </Modal>
+      </React.Fragment>
+    );
+  }
+}
+
+const mapStateToProps = (state) => ({
+  isFeatureToggledOn: hasUiOption('snippet')(state)
```
Comment (ischer) [inline src/pages/templates/components/ImportSnippetLink.js]: tiny typo here - `snippet` -> `snippets`

File: src/pages/templates/EditPage.js (hunk 5 above)
Comment (ischer) [reply src/pages/templates/EditPage.js]: https://github.com/SparkPost/2web2ui/pull/694/files#r233250407 - tiny typo on the option check

File: src/pages/templates/components/ImportSnippetPanel.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+      return <PanelLoading />;
+    }
+
+    return (
+      <Panel
+        title="Import snippets in this template"
+        actions={[{
+          component: ExternalLink,
+          content: 'Learn More',
+          to: LINKS.SNIPPET_SUBSTITUTION_REFERENCE
```
Comment (ischer) [inline src/pages/templates/components/ImportSnippetPanel.js]: not sure if it matters, but this link ```(https://developers.sparkpost.com/api/substitutions-reference/#header-snippets)``` redirects to ```https://developers.sparkpost.com/api/template-language/#header-snippets```. should this be updated?

Comment (bkemper) [reply src/pages/templates/components/ImportSnippetPanel.js]: I just modeled it off [this link](https://github.com/SparkPost/developers.sparkpost.com/pull/247/files#diff-f4046802ad48df82cf6d2c43d51df5d1R37) which should load [this section](https://github.com/SparkPost/developers.sparkpost.com/pull/247/files#diff-f4046802ad48df82cf6d2c43d51df5d1R40).

File: src/pages/templates/components/ImportSnippetPanel.js (hunk 2 above)
Comment (bkemper) [reply src/pages/templates/components/ImportSnippetPanel.js]: Addressed... ef072d4f0

**Comment by bkemper** [comment]
@ischer All remaining feedback has been addressed.  Ready for final final review.

**Comment by bkemper** [comment]
@ischer Rebased!

**Pull request merged**